### PR TITLE
feat: Add adaptive cutoff neighbor-list factory with count-based dense/cell-list policy

### PR DESCRIPTION
**Pull request by n-gao**
### Motivation

Previously, every simulation state class (`MCMCState`, `LjMdState`, `MlffMdState`, `RelaxLjState`, etc.) was required to implement a `neighborlist(cutoffs)` method that hardcoded `DenseNearestNeighborList.from_state`. This made it impossible to swap in a different neighbor-list implementation (e.g., `CellListNeighborList`) without modifying each state class, and it coupled potential factories directly to the state's method dispatch.

### Changes

**`adaptive.py` (new module)**

Introduces `adaptive_cutoff_neighborlist_from_state`, a factory function that selects between `DenseNearestNeighborList` and `CellListNeighborList` at construction time using a count-based `CutoffNeighborListPolicy`. The policy avoids inspecting cutoff values or cell geometry (which are unreliable at construction time) and instead uses average particles-per-system as a stable heuristic. The default threshold is 10,000 particles per system. The `CutoffNeighborListStrategy` enum (`AUTO`, `DENSE`, `CELL_LIST`) allows the policy to be overridden for debugging or benchmarking without changing call sites.

**`NeighborListFactory` protocol**

A new `NeighborListFactory[State]` protocol is added to `protocol.py`, formalizing the `(state, cutoffs) -> NeighborList[Literal[2]]` signature. All potential `make_*_from_state` functions (`make_lennard_jones_from_state`, `make_ewald_from_state`, `make_coulomb_vacuum_from_state`, `make_blocking_spheres_from_state`, `make_local_mliap_from_state`, `make_tojaxed_from_state`, `make_torch_mliap_from_state`) now accept a `neighborlist_factory` keyword argument defaulting to `adaptive_cutoff_neighborlist_from_state`.

**State protocol cleanup**

The `neighborlist(cutoffs)` method is removed from all state protocols (`IsLJState`, `IsEwaldState`, `IsCoulombVacuumState`, `IsBlockingSpheresState`, `IsLocalMLIAPState`, `IsTojaxedState`, `IsTorchMliapState`) and from all concrete state classes. These protocols now extend `IsAdaptiveCutoffNeighborListState` instead, requiring only `neighborlist_params`, `particles.size`, and `systems.size`.

**`ReduceCompactor` and `MaskOnlyCompactor`**

Both compactors are made generic over the edge dimension `D` and now explicitly implement `Compactor[D]`, replacing the previously hardcoded `Literal[2]` signatures.

**Tests**

New tests in `test_neighborlist.py` cover `CutoffNeighborListPolicy.choose` across threshold boundaries, zero-system edge cases, and custom thresholds, as well as `adaptive_cutoff_neighborlist_from_state` returning the correct concrete type under `AUTO`, forced `DENSE`, and forced `CELL_LIST` strategies. The potential update tests inject a custom `neighborlist_factory` to avoid cell-list sizing constraints in small test fixtures.

### CODE REVIEW

**Comment by n-gao** [comment]
* **#171** <a href="https://app.graphite.com/github/pr/cusp-ai-oss/kUPS/171?utm_source=stack-comment-icon" target="_blank">Graphite</a>: 2  dependent PRs ([#173](https://github.com/cusp-ai-oss/kUPS/pull/173) <a href="https://app.graphite.com/github/pr/cusp-ai-oss/kUPS/173?utm_source=stack-comment-icon" target="_blank">Graphite</a>, [#178](https://github.com/cusp-ai-oss/kUPS/pull/178) <a href="https://app.graphite.com/github/pr/cusp-ai-oss/kUPS/178?utm_source=stack-comment-icon" target="_blank">Graphite</a>)
* **#170** <a href="https://app.graphite.com/github/pr/cusp-ai-oss/kUPS/170?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#169** <a href="https://app.graphite.com/github/pr/cusp-ai-oss/kUPS/169?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#168** <a href="https://app.graphite.com/github/pr/cusp-ai-oss/kUPS/168?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#167** <a href="https://app.graphite.com/github/pr/cusp-ai-oss/kUPS/167?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#166** <a href="https://app.graphite.com/github/pr/cusp-ai-oss/kUPS/166?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#165** <a href="https://app.graphite.com/github/pr/cusp-ai-oss/kUPS/165?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#164** <a href="https://app.graphite.com/github/pr/cusp-ai-oss/kUPS/164?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#155** <a href="https://app.graphite.com/github/pr/cusp-ai-oss/kUPS/155?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#154** <a href="https://app.graphite.com/github/pr/cusp-ai-oss/kUPS/154?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#153** <a href="https://app.graphite.com/github/pr/cusp-ai-oss/kUPS/153?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#152** <a href="https://app.graphite.com/github/pr/cusp-ai-oss/kUPS/152?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#151** <a href="https://app.graphite.com/github/pr/cusp-ai-oss/kUPS/151?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#150** <a href="https://app.graphite.com/github/pr/cusp-ai-oss/kUPS/150?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#149** <a href="https://app.graphite.com/github/pr/cusp-ai-oss/kUPS/149?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#148** <a href="https://app.graphite.com/github/pr/cusp-ai-oss/kUPS/148?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈 <a href="https://app.graphite.com/github/pr/cusp-ai-oss/kUPS/148?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* **#147** <a href="https://app.graphite.com/github/pr/cusp-ai-oss/kUPS/147?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#146** <a href="https://app.graphite.com/github/pr/cusp-ai-oss/kUPS/146?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#145** <a href="https://app.graphite.com/github/pr/cusp-ai-oss/kUPS/145?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#144** <a href="https://app.graphite.com/github/pr/cusp-ai-oss/kUPS/144?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#143** <a href="https://app.graphite.com/github/pr/cusp-ai-oss/kUPS/143?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#142** <a href="https://app.graphite.com/github/pr/cusp-ai-oss/kUPS/142?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#141** <a href="https://app.graphite.com/github/pr/cusp-ai-oss/kUPS/141?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#138** <a href="https://app.graphite.com/github/pr/cusp-ai-oss/kUPS/138?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#137** <a href="https://app.graphite.com/github/pr/cusp-ai-oss/kUPS/137?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#136** <a href="https://app.graphite.com/github/pr/cusp-ai-oss/kUPS/136?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#135** <a href="https://app.graphite.com/github/pr/cusp-ai-oss/kUPS/135?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#122** <a href="https://app.graphite.com/github/pr/cusp-ai-oss/kUPS/122?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#121** <a href="https://app.graphite.com/github/pr/cusp-ai-oss/kUPS/121?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#134** <a href="https://app.graphite.com/github/pr/cusp-ai-oss/kUPS/134?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#133** <a href="https://app.graphite.com/github/pr/cusp-ai-oss/kUPS/133?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#132** <a href="https://app.graphite.com/github/pr/cusp-ai-oss/kUPS/132?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#131** <a href="https://app.graphite.com/github/pr/cusp-ai-oss/kUPS/131?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#120** <a href="https://app.graphite.com/github/pr/cusp-ai-oss/kUPS/120?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#119** <a href="https://app.graphite.com/github/pr/cusp-ai-oss/kUPS/119?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#118** <a href="https://app.graphite.com/github/pr/cusp-ai-oss/kUPS/118?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* `main`


<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.

File: src/kups/core/neighborlist/adaptive.py [hunk 1]
```diff
@@ -0,0 +1,140 @@
...
+from kups.core.neighborlist.dense import DenseNearestNeighborList
+from kups.core.neighborlist.types import (
+    IsUniversalNeighborlistParams,
+    NeighborList,
+)
+from kups.core.typing import SystemId
+from kups.core.utils.jax import dataclass, field
+
+
+class CutoffNeighborListStrategy(StrEnum):
```
Comment (jonkhler) [inline src/kups/core/neighborlist/adaptive.py:38]: why not inject behavior with functions as we do elswhere? enum strategy seems like a smell to me

**Review by jonkhler — CHANGES_REQUESTED** [review]
Not happy about the enum as this is very different style as we use throughout the code base.  
  
Also more of a question that a change request: are we buying more and more technical debts due to increasing complexity with factories, strategies, adapters etc. etc.?

**Comment by n-gao** [comment]
The goal of this being is that it is a lightweight wrapper that dispatches to a reasonable choice.  
  
As I see it we can do:  
1\. This approach.  
2\. Always be explicit about the neighborlist algorithm (we then have to pick a sensible default, which is what this PR tries to solve/avoid?)  
3\.  We could have a defined NeighborlistPolicy class. Each neighborlists dispatches a priority via an integer on their preference to be selected depending on static variables. Then the adaptive neighborlist picks the algorithm with the highest priority.  
  
While I see 3 being the most general solution, it would also require a "default set" of policies and seems "overengineered" for the problem at hand?  
  
Do you have other suggestions?

**Review by jonkhler — APPROVED** [review]
ok let's do it like this for now but be careful with further spec drift

**Comment by n-gao** [comment]
### Merge activity

* **Jun 16, 1:22 PM UTC**: @n-gao merged this pull request with [Graphite](https://app.graphite.com/github/pr/cusp-ai-oss/kUPS/148).
